### PR TITLE
Better align shutdown semantics of testing event loops

### PR DESCRIPTION
**Pull request by simonjbeaumont**
### Motivation:

The two testing event loops—`EmbeddedEventLoop` and `NIOAsyncTestingEventLoop`—have different semantics for outstanding work during shutdown, which are both different from the production `SelectableEventLoop`, specifically with newly scheduled tasks that result from running existing scheduled work at the time of shutdown.

There are three axes to consider:

1. Scheduled tasks that are at or past their deadline.
2. Scheduled tasks with a deadline in the future.
3. Newly scheduled work that result from either running or cancelling (1) and (2).

| | (1) | (2) | (3) |
|-|-|-|-|
| `SelectableEventLoop` | Run | Cancel | Quiesce over 1000 ticks, repeatedly run resulting (1) and cancel resulting (2) |
| `EmbeddedEventLoop` | Run | Cancel | Cancel resulting (1) and resulting (2) |
| `NIOAsyncTestingEventLoop` | Run | Run | Run resulting (1) and resulting (2) |

Note that `NIOAsyncTestingEventLoop` may never terminate because of this.

### Modifications:

This PR aligns `EmbeddedEventLoop` and `NIOTestingEventLoop` and makes them more similar to `SelectableEventLoop` and less surprising semantics.

### Result:

| | (1) | (2) | (3) |
|-|-|-|-|
| `SelectableEventLoop` | Run | Cancel | Quiesce over 1000 ticks, repeatedly run resulting (1) and cancel resulting (2) |
| `EmbeddedEventLoop` | Run | Cancel | Cancel resulting (1) and resulting (2) |
| `NIOAsyncTestingEventLoop` | Run | Cancel | Cancel resulting (1) and resulting (2) |

### Future work:

Given both the `EmbeddedEventLoop` and `NIOAsyncTestingEventLoop` are used in tests of NIO applications that run with `SelectableEventLoop` in production, it might be worth extending both to also support quiescing to give a more representative shutdown behaviour in tests.

### CODE REVIEW

File: Sources/NIOEmbedded/AsyncTestingEventLoop.swift [hunk 1]
```diff
@@ -95,6 +95,10 @@ public final class NIOAsyncTestingEventLoop: EventLoop, @unchecked Sendable {
     /// The queue on which we run all our operations.
     private let queue = DispatchQueue(label: "io.swiftnio.AsyncEmbeddedEventLoop")
 
+    private enum State: Int, AtomicValue { case open, closing, closed }
+    private let _state = ManagedAtomic(State.open)
+    private var state: State { self._state.load(ordering: .relaxed) }
```
Comment (Lukasa) [inline Sources/NIOEmbedded/AsyncTestingEventLoop.swift:100]: The atomic ordering here seems likely to be logically wrong. We should probably force ourselves to express it at all points.

File: Sources/NIOEmbedded/AsyncTestingEventLoop.swift [hunk 2]
```diff
@@ -150,6 +154,15 @@ public final class NIOAsyncTestingEventLoop: EventLoop, @unchecked Sendable {
         let promise: EventLoopPromise<T> = self.makePromise()
         let taskID = self.scheduledTaskCounter.loadThenWrappingIncrement(ordering: .relaxed)
 
+        switch self.state {
```
Comment (Lukasa) [inline Sources/NIOEmbedded/AsyncTestingEventLoop.swift:157]: The load here should be `acquiring`.

File: Sources/NIOEmbedded/AsyncTestingEventLoop.swift [hunk 3]
```diff
@@ -324,9 +334,11 @@ public final class NIOAsyncTestingEventLoop: EventLoop, @unchecked Sendable {
 
     /// The concurrency-aware equivalent of `shutdownGracefully(queue:_:)`.
     public func shutdownGracefully() async {
-        await withCheckedContinuation { (continuation: CheckedContinuation<Void, Never>) in
+        await withCheckedContinuation { continuation in
+            self._state.store(.closing, ordering: .relaxed)
```
Comment (Lukasa) [inline Sources/NIOEmbedded/AsyncTestingEventLoop.swift:338]: These stores should be `releasing`.

File: Sources/NIOEmbedded/AsyncTestingEventLoop.swift (hunk 1 above)
Comment (simonjbeaumont) [reply Sources/NIOEmbedded/AsyncTestingEventLoop.swift]: Sounds good. I saw a bunch of `.relaxed` in this implementation already but I think they're probably OK. I realise that this isn't appropriate here now. Updated.

**Review by Lukasa — APPROVED** [review]

**Comment by simonjbeaumont** [comment]
@FranzBusch I just disabled automerge on this one because I wanted you to +1 the updates to the threshold JSON. I'll have another PR with update to the dev script for making this easier.

**Review by FranzBusch — APPROVED** [review]

**Pull request merged**